src/posts/2022-11-03-constellation/index.md

```diff
@@ -0,0 +1,148 @@
+4. Cilium-agent watches and consumes all CiliumEndpointSlices
+5. Cilium-agent writes a new entry in IPCache
+
+From this flow, it is obvious that the IPCache is only eventually consistent, and we should be able to reproduce the issue by disabling the Cilium operator.
+
+To deterministically reproduce the issue, we can adapt the WireGuard integration test in [`test/k8s/datapath_configuration.go`](https://github.com/edgelesssys/cilium/blob/d465085df155502267589f782801c63f3d08dd40/test/k8s/datapath_configuration.go#L545).  In this test, the execution is explicitly halted until the endpoint has propagated:
+
+```golang
+// Due to IPCache update delays, it can take up to a few seconds
+// before both nodes have added the new pod IPs to their allowedIPs
+// list, which can cause flakes in CI. Therefore wait for the
+// IPs to be present on both nodes before performing the test
+```
+
+Using this test, we scale the Cilium-operator deployment to 0 before deploying the test daemonset. If we now remove the check for the updated WireGuard peers mentioned above, the tests fail since the new endpoints can’t properly propagate to the IPCache and unencrypted traffic is leaked.
+
+In summary, when network traffic is sent to newly created Kubernetes Pods, a short delay in the propagation of routing information can lead to Cilium sending that traffic unencrypted over the wire.
+ 
+### The strict mode solution
+
+Now that we [have](https://github.com/edgelesssys/cilium/blob/d465085df155502267589f782801c63f3d08dd40/test/k8s/datapath_configuration.go#L631) a test case, we can begin to develop a solution. We call it the strict mode.
+
+With Constellation we are in the unique position that we control the whole underlying stack, including the Kubernetes and Cilium bootstrapping process. Therefore, we can automatically deploy and configure Cilium with our strict mode.
+
+We define the following constraints for our strict mode:
+
+- IPv4 only (Pod network commonly IPv4 only)
+- PodCIDR must be known beforehand
+- Both vxlan (used for AWS/Azure) and native routing (used for GCP) must be supported
+
+While the IPv4 limitation can be easily solved later, the PodCIDR must be static and known before deploying Cilium. Support for a dynamic PodCIDR is a complex feature to implement, because of the same eventual consistency problems we are trying to solve in the first place. Since we support Cilium’s tunneling and direct routing capabilities, the strict mode must also work with both. Note that the test we adapted already covers both scenarios.
+
+> A strict mode for AWS and Azure with direct routing poses some challenges. Since Pod IPs are allocated out of the same subnet as Node IPs, we cannot filter by subnet. Moreover, we cannot rely on the IPCache to add something like an allow-node rule to the filter. The Node could have been terminated and its IP re-used as a Pod IP while the IP is still associated with the Node in the IPCache. For that reason, we’re using vxlan.
```

Comment:
> For that reason, we’re using vxlan.

Does that mean that you are setting Node IPs subnet as the strict one?

Reply:
@gandro is saying that the usage of the vxlan means that you are not using clouds native routing, and thus ClusterPool IPAM is used which does not run into the problem.

Reply:
I'm sorry, if I don't fully get your question, but we are using vxlan for AWS and Azure as a means to have the Pods in their own CIDR, which we can filter. On GCP we can still use direct routing since their AliasIPRange still allows us to keep the PodCIDR separate. So we always set the PodCIDR as the strict one. 
In line 87 we wanted to clarify why we take those different approaches aka. why a strict mode + direct routing on AWS/Azure is difficult to implement.